Emit event when changing total locked value in pallet-balances

sea212 opened the pull request:
Fixes #12276 

This PR adds two events to the balances pallet, namely:
```rust
/// Some free balance was locked out from specific operations.
Locked { who: T::AccountId, amount: T::Balance, reason: Reasons },
/// Some balance that was locked out from specific operations was freed.
Unlocked { who: T::AccountId, amount: T::Balance, reason: Reasons },
```

Since locks in the same category of `Reasons` overlay each other, the design decision was made to include only the changes of total lock value in the `amount` fields of the respective events. Examples:
```rust
Locks = [10/MISC]          --> [15/MISC]              // emits Locked(..., amount=5, Reasons::Misc)
Locks = [15/MISC]          --> [15/MISC, 20/MISC]     // emits Locked(..., amount=5, Reasons::Misc)
Locks = [15/MISC, 20/MISC] --> [17/MISC, 20/MISC]     // emits nothing
Locks = [17/MISC, 20/MISC] --> [20/MISC]              // emits nothing
Locks = [20/MISC]          --> [10/MISC]              // emits Unlocked(..., amount=10, Reasons::Misc)
Locks = [10/MISC]          --> [10/MISC, 15/FEE]      // emits Locked(..., amount=15, Reasons::Fee)
Locks = [10/MISC, 15/FEE]  --> [20/MISC, 20/FEE]      // emits Locked(..., amount=10, Reasons::Misc) and
                                                      // emits Locked(..., amount=5, Reasons::Fee)
// etc.
```

There are also alternative approaches that could be applied:
1. Add `total_locked` field to the respective events.
2. Emit lock event that only contains `total_locked` instead of the `amount` that was added to or removed from the total lock value.

Some questions in regards to the actual changes in this PR:
1. Currently `Reasons::All` is not used in the event. There is a corner case when a lock is set with `Reasons::All` that implies the same relative change of lock amounts for both `Reasons::Misc` and `Reasons::Fee` (all currently available reasons). In that case the event could emit `Reasons::All`. I decided against adding this because in terms of implementation effort this does not scale well at all. Instead, two distinct events are now emitted.
2. The PR contains this logic: `if a > b then a - b` and `if a < b then b - a`. Is it okay to use non-saturating math here or is using saturating math only enforced meanwhile?

sea212 committed:
Emit Locked/Unlocked events

sea212 committed:
Implement lock event tests

sea212 committed:
Adhere to style guide

kianenigma commented on frame/balances/src/lib.rs:
```diff
@@ -994,6 +1009,19 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+
+		let deposit_lock_event = |prev: T::Balance, after: T::Balance, reason: Reasons| {
+			if prev < after {
+				let amount = after - prev;
```
```suggestion
				let amount = after.saturaitng_sub(prev);
```

ggwpez replied:
Or

```rust
if let Some(amount) = after.checked_sub(prev) {
```

kianenigma commented on frame/balances/src/lib.rs:
```diff
@@ -994,6 +1009,19 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+				let amount = after - prev;
+				Self::deposit_event(Event::Locked { who: who.clone(), amount, reason });
+			} else if prev > after {
+				let amount = prev - after;
```
```suggestion
				let amount = prev.saturaitng_sub(after);
```

sea212 committed:
Use saturating math

Co-authored-by: Kian Paimani <5588131+kianenigma@users.noreply.github.com>

sea212 committed:
Fix typo

melekes reviewed (approved):
👍 (code-wise; don't know enough to reason about business logic)

ggwpez reviewed (approved):
Putting it up for audit :+1:

ggwpez commented:
bot rebase

ggwpez commented on frame/balances/src/lib.rs:
```diff
@@ -991,6 +1006,19 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+
+		let deposit_lock_event = |prev: T::Balance, after: T::Balance, reason: Reasons| {
+			if prev < after {
+				let amount = after.saturating_sub(prev);
```
>Is it okay to use non-saturating math here or is using saturating math only enforced meanwhile?

Would be okay here. But since you already have it - its doing no harm.

ggwpez commented:
>Emit lock event that only contains total_locked instead of the amount that was added to or removed from the total lock value.

I think both is fine. In both cases you either dont have the *diff* or not the *total* and putting both in the event seems like overkill.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2214503

abebeos commented:
Project status is still `Status: Needs Audit`, there should  be another status like `Audited` or `Mergable`.

(Otherwise merge should be done right after audit)

bkchr commented:
bot rebase

bkchr commented:
@sea212 please merge master.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2493772

sea212 commented:
> @sea212 please merge master.

Rebased the branch onto main and squashed some commits together.

You can [compare](https://github.com/paritytech/substrate/compare/master...sea212:substrate:1423d497c78d96cda2c1c58adf71cb4482d122b0) that only the commits from Substrate were rebased unmodified onto main and you can [compare](https://github.com/sea212/substrate/compare/d4f36848f79b64aedd064117aca7103a210d15c5...1423d497c78d96cda2c1c58adf71cb4482d122b0) that the changes in this PR are equal pre and post rebase.

bkchr commented:
@sea212 can you please fix the CI errors:

```
--- STDERR:              pallet-offences-benchmarking bench_report_offence_im_online ---
thread 'bench_report_offence_im_online' panicked at 'Mismatching events.', frame/offences/benchmarking/src/lib.rs:276:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/panicking.rs:65:14
   2: core::ops::function::FnOnce::call_once{{vtable.shim}}
   3: pallet_offences_benchmarking::Pallet<T>::test_benchmark_report_offence_babe::{{closure}}
   4: pallet_offences_benchmarking::Pallet<T>::test_benchmark_report_offence_im_online
   5: std::thread::local::LocalKey<T>::with
   6: core::ops::function::FnOnce::call_once
   7: core::ops::function::FnOnce::call_once
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/ops/function.rs:251:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'bench_report_offence_im_online' panicked at 'report_offence_im_online: Any { .. }', frame/offences/benchmarking/src/lib.rs:286:1
stack backtrace:
```

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2585423

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2586049

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2586050

sea212 committed:
Merge branch 'master' into balances-add-lock-events

sea212 committed:
Emit event on change locks and add tests

sea212 committed:
Adjust event docstring

sea212 commented:
Since some big changes in `pallet-balances` were introduced on main, this PR probably needs a review again. Essentially everything stayed the same, but now the lock `Reason` is ignored in the `Locked/Unlocked` events. In addition to that, the tests were adjusted to ignore the lock `Reason` as well. Those conditions are tested now:
```rust
Locks = []        --> [10]      // emits Locked(..., amount=10)
Locks = [10]      --> [15]      // emits Locked(..., amount=5)
Locks = [15]      --> [15, 20]  // emits Locked(..., amount=5)
Locks = [15, 20]  --> [17, 20]  // emits nothing
Locks = [17, 20]  --> [17, 15]  // emits Unlocked(..., amount=3)
Locks = [17, 15]  --> [15]      // emits Unlocked(..., amount=2)
Locks = [15]      --> []        // emits Unlocked(..., amount=15)
```

For the sake of clarity and convenience to review, decided to create a second PR that handles the events for the new `freeze` and `thaw` calls that deprecate the locks.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2586472

sea212 commented:
Does anyone know what is going on here in regards to the CI? If I run the command locally all tests pass. The changes in this PR are even not related to the errors that occur.

bkchr commented:
bot merge

bkchr commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into balances-add-lock-events

bkchr commented:
bot merge

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2589497

ggwpez commented:
Master is red now, but there were internal CI problems. Will fix.
